File: lclang/src/main/java/lclang/statements/expressions/ArrayExpression.java
package lclang.statements.expressions;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.exceptions.LCLangTypeErrorException;
import lclang.libs.lang.classes.ArrayClass;
import lclang.utils.LinkUtils;

import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

public class ArrayExpression extends Expression {
    public final List<Expression> array;

    public ArrayExpression(List<Expression> array) {
        super(0);
        this.array = array;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor executor) throws LCLangRuntimeException {
        final List<Link> values = LinkUtils.linksFromExpressions(prevCaller, executor, array);
        AtomicReference<ArrayClass> arrayClass = new AtomicReference<>(null);
        return new Link(caller -> {
            if(arrayClass.get()==null) arrayClass.set(new ArrayClass(
                    LinkUtils.classesFromLinks(caller, values)
            ));

            return arrayClass.get();
        }, (setCaller, clazz) -> {
            if(clazz instanceof ArrayClass) {
                ArrayClass otherArray = (ArrayClass) clazz;
                for (int i = 0; i < array.size(); i++) {
                    values.get(i).set.invoke(setCaller, otherArray.get(i));
                }
            }else throw new LCLangTypeErrorException("Value is not array", setCaller);
        }, Link.State.NOTHING);
    }
}

File: lclang/src/main/kotlin/lclang/statements/ArrayExpression.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.Value
import lclang.exceptions.TypeErrorException
import lclang.libs.lang.ArrayClass

class ArrayExpression(
    val array: List<Expression>,
    line: Int,
    column: Int
): Expression(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        val caller = getCaller(prevCaller)
        val arrayClass = ArrayClass(array.map { it.visit(caller, visitor) })
        val value = arrayClass.asValue()
        value.set = { c, it ->
            val otherArray = it.get(c)
            if(otherArray !is ArrayClass) throw TypeErrorException("Value is not array", c)

            for((i, expression) in array.withIndex()){
                expression.visit(c, visitor).set(c, otherArray[i])
            }
        }

        return value
    }
}